rust/lance-encoding/src/encodings/physical/bitpack_fastlanes.rs

```diff
@@ -0,0 +1,1547 @@
+                            DataType::UInt64 | DataType::Int64 => encoded_values = encode_fixed_width!(self, unpacked, u64, buffer_index)?,
+                            _ => unreachable!("BitpackedForNonNegArrayEncoder only supports data types of UInt8, Int8, UInt16, Int16, UInt32, Int32, UInt64, Int64"),
+                        }
+                    }
+                    _ => {
+                        return Err(Error::InvalidInput {
+                            source: "Bitpacking only supports fixed width data blocks or a nullable data block with fixed width data block inside or a all null data block".into(),
+                            location: location!(),
+                        });
+                    }
+                }
+                let encoding =
+                    ProtobufUtils::basic_some_null_encoding(validity_desc, encoded_values.encoding);
+                let encoded = DataBlock::Nullable(NullableDataBlock {
+                    data: Box::new(encoded_values.data),
+                    nulls: nullable.nulls,
+                });
+                Ok(EncodedArray {
+                    data: encoded,
+                    encoding,
+                })
+            }
+            _ => {
+                Err(Error::InvalidInput {
+                    source: "Bitpacking only supports fixed width data blocks or a nullable data block with fixed width data block inside or a all null data block".into(),
+                    location: location!(),
+                })
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct BitpackedForNonNegScheduler {
+    compressed_bit_width: u64,
+    uncompressed_bits_per_value: u64,
+    buffer_offset: u64,
+}
+
+fn locate_chunk_start(scheduler: &BitpackedForNonNegScheduler, relative_row_num: u64) -> u64 {
+    let chunk_size = ELEMS_PER_CHUNK * scheduler.compressed_bit_width / 8;
+    relative_row_num / ELEMS_PER_CHUNK * chunk_size
+}
+
+fn locate_chunk_end(scheduler: &BitpackedForNonNegScheduler, relative_row_num: u64) -> u64 {
+    let chunk_size = ELEMS_PER_CHUNK * scheduler.compressed_bit_width / 8;
+    relative_row_num / ELEMS_PER_CHUNK * chunk_size + chunk_size
+}
+
+impl BitpackedForNonNegScheduler {
+    pub fn new(
+        compressed_bit_width: u64,
+        uncompressed_bits_per_value: u64,
+        buffer_offset: u64,
+    ) -> Self {
+        Self {
+            compressed_bit_width,
+            uncompressed_bits_per_value,
+            buffer_offset,
+        }
+    }
+}
+
+impl PageScheduler for BitpackedForNonNegScheduler {
+    fn schedule_ranges(
+        &self,
+        ranges: &[std::ops::Range<u64>],
+        scheduler: &Arc<dyn crate::EncodingsIo>,
+        top_level_row: u64,
+    ) -> BoxFuture<'static, Result<Box<dyn PrimitivePageDecoder>>> {
+        // can we get here?
+        if ranges.is_empty() {
+            panic!("cannot schedule empty ranges");
+        }
```

Comment:
shouldn't be able to.  Feel free to make a (not debug) assert

Reply:
fixed